### PR TITLE
newapp - retain ordering of env vars

### PR DESCRIPTION
**Pull request by csrwng**
Changes the underlying structure of the Environment type to
a slice instead of a map.

Fixes bz 1430277

### CODE REVIEW

**Comment by csrwng** [comment]
@openshift/sig-developer-experience ptal
/unassign @knobunc

File: pkg/generate/app/cmd/newapp_test.go [hunk 1]
```diff
@@ -141,7 +141,7 @@ func TestValidate(t *testing.T) {
 		if len(env) != len(c.env) {
 			t.Errorf("%s: Environment variables don't match. Expected: %v, Got: %v", n, c.env, env)
 		}
-		for e, v := range env {
+		for e, v := range env.Map() {
```
Comment (jim-minter) [inline pkg/generate/app/cmd/newapp_test.go]: not sure this is right - shouldn't c.{env,buildEnv,parms} be redefined as non-maps to test ordering?

File: pkg/generate/app/env.go [hunk 2]
```diff
@@ -137,11 +178,11 @@ func LoadEnvironmentFile(filename string, stdin io.Reader) (Environment, error)
 	errorFilename := filename
 
 	if filename == "-" && stdin != nil {
-		//once https://github.com/joho/godotenv/pull/20 is merged we can get rid of using tempfile
 		temp, err := ioutil.TempFile("", "origin-env-stdin")
 		if err != nil {
 			return nil, fmt.Errorf("Cannot create temporary file: %s", err)
 		}
+		defer temp.Close()
```
Comment (jim-minter) [inline pkg/generate/app/env.go]: remove temp.Close() below

File: pkg/generate/app/env.go [hunk 3]
```diff
@@ -158,19 +199,21 @@ func LoadEnvironmentFile(filename string, stdin io.Reader) (Environment, error)
 	if info, err := os.Stat(filename); err == nil && info.IsDir() {
 		return nil, fmt.Errorf("Cannot read variables from %q: is a directory", filename)
 	} else if err != nil {
-		return nil, fmt.Errorf("Cannot stat %q: %s", filename, err)
+		return nil, fmt.Errorf("Cannot stat %q: %v", filename, err)
 	}
 
 	env, err := godotenv.Read(filename)
```
Comment (jim-minter) [inline pkg/generate/app/env.go]: godotenv.Read() returns a map, which will screw up the k/v ordering when reading in the file.  I don't particularly mind, but perhaps the BZ submitter will...
This is also the only case where godotenv is used within origin.  I'd be inclined to blitz the dependency and TempFile nonsense above and read the file in properly by steam.

File: pkg/ipfailover/keepalived/generator.go [hunk 4]
```diff
@@ -39,8 +36,8 @@ func generateEnvEntries(name string, options *ipfailover.IPFailoverConfigCmdOpti
 		"OPENSHIFT_HA_PREEMPTION":        options.Preemption,
 		"OPENSHIFT_HA_CHECK_INTERVAL":    interval,
 		// "OPENSHIFT_HA_UNICAST_PEERS":     "127.0.0.1",
-	})
-	return env
+	}
+	return app.NewEnvironmentFromMap(env)
```
Comment (jim-minter) [inline pkg/ipfailover/keepalived/generator.go]: my inclination would be to factor out usage of app.Environment from outside its home territory.

**Comment by csrwng** [comment]
Closing this for now

**Pull request closed without merging**